<!DOCTYPE html>
<html lang="en">

  <head>

    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
    <meta name="description" content="">
    <meta name="author" content="">

    <title>Wright's Fst</title>

    <!-- Bootstrap core CSS -->
    <link href="/vendor/bootstrap/css/bootstrap.min.css" rel="stylesheet">

    <!-- Custom styles for this template -->
    <link href="/css/1-col-portfolio.css" rel="stylesheet">

  </head>

  <body>
  <!-- Navigation -->
    <nav class="navbar navbar-expand-lg navbar-dark bg-dark fixed-top">
      <div class="container">
        <a class="navbar-brand" href="#">Fst-Exercises</a>
        <button class="navbar-toggler" type="button" data-toggle="collapse" data-target="#navbarResponsive" aria-controls="navbarResponsive" aria-expanded="false" aria-label="Toggle navigation">
          <span class="navbar-toggler-icon"></span>
        </button>
        <div class="collapse navbar-collapse" id="navbarResponsive">
          <ul class="navbar-nav ml-auto">
            <li class="nav-item active">
              <a class="nav-link" href="/compGen8900.html">Gen-8900
                <span class="sr-only">(current)</span>
              </a>
	    </li>
	    <li class="nat-item active">
	      <a class="nav-link" href="/gen8900/exercises/hwe.html">HWE-Exercises
	        <span class="sr-only">(current)</span>
	      </a>
	    </li>
	    <li class="nat-item active">
	      <a class="nav-link" href="/gen8900/exercises/ld-1.html">LD-Part I
		<span class="sr-only">(current)</span>
	      </a>
	      </li>
	  </ul>
	 </div>
	</div>
	</nav>

	<!--Page Content -->
	<section id="vcf-exercises">
	<div class="container">
	  <div class="row">
	  <div-class="col-md-12">
	    <h1>Population Differentiation with Fst</h1>
<h3>Before Starting:</h3>
<p>Remember to either load my <a href="https://gist.github.com/eacooper400/c954c757db02eb6c4ccfae1aa090658c">CompGen_Fxns17.R</a> file or copy and paste the <code>read.vcf</code> function into your code.</p>
<p>Then, load the sample data file for this week: sampleData_4_Fst.vcf</p>
<h3 id="1gettheexpectedhetvaluesforeachpopulation">1. Get the EXPECTED HET values for each population</h3>
<h4 id="partasetuptheemptyresultstable">Part A: Set up the Empty Results Table</h4>
<p>This table should have:</p>
<ul>
<li>The same number of rows as <code>my.data</code></li>
<li><em>At least</em> 9 columns:</li>
</ul>
<p style="padding-left: 60px;"><strong>p</strong>: The frequency of allele (A) in the total population</p>
<p style="padding-left: 60px;"><strong>n</strong>: The number of non-missing individuals in the total population.</p>
<p style="padding-left: 60px;"><strong>Hexp</strong>: The expected heterozygosity for the total population (frequency, not counts!!!).</p>
<p style="padding-left: 60px;"><strong>p1</strong>: The frequency of allele (A) in sub-population 1.</p>
<p style="padding-left: 60px;"><strong>n1</strong>: The number of (non-missing) individuals in sub-pop 1</p>
<p style="padding-left: 60px;"><strong>Hexp1</strong> The expected heterozygosity in sub-pop 1</p>
<p style="padding-left: 60px;"><strong>p2</strong>: The frequency of allele (A) is sub-population 2.</p>
<p style="padding-left: 60px;"><strong>n2</strong>: The number of non-missing individuals in sub-pop 2.</p>
<p style="padding-left: 60px;"><strong>Hexp2</strong>: The expected heterozygosity for sub-pop 2.</p>
<p><em>Note: Feel free to include additional columns in your table, such as columns for the observed genotype counts, if you want them there to help with the calculations.</em></p>
<h4 id="partbcalculateexpaaforthetotalpopulation">Part B: Calculate Het<sub>exp</sub> for the Total population</h4>
<p>This calculation is exactly the same as the calculation we did for the Hardy-Weinberg exercises last week, so you should be able to re-use your code and fill in the appropriate values of your new results table. (Although note again that this time you <span style="text-decoration: underline;">do not</span> need the expected <strong>count</strong> of Aa individuals, just their expected <strong>frequency</strong> (i.e. just 2pq))</p>
<h4 id="partccalculateexpaaforthe2subpopulations">Part C: Calculate H<sub>exp</sub> for the 2 Sub-Populations</h4>
<p>The first step in this part of the exercise is to figure out which columns of <code>my.data</code> correspond to each sub-population.</p>
<p>In this file, the samples from the first sub-population (<em>Geospiza fuliginosa</em>) all have names that start with "FU_".</p>
<p>The samples from the second sub-population (<em>Geospiza magnirostris</em>) all have names that start with "MA_".</p>
<p>You can extract the column names from <code>my.data</code> like this:</p>
<pre><code>colnames(my.data)
</code></pre>
<p>You can use <code>grep()</code> on the column names to get the index of the columns that match a particular pattern.</p>
<p>Then, you can use the square bracket indexing notation to pull out only those columns from the <code>my.data</code> table:</p>
<pre><code>my.data[ , (<span style="color: #ff0000;">your grep results</span>)]
</code></pre>
<p>Once you're able to get only the parts of the table that correspond to one sub-population, you can use the exact same methods for calculating H<sub>exp</sub> that you used for the total population.</p>
<h3 id="2useexpectedhetvaluestocalculatefst">2. Use Expected Het. values to Calculate Fst</h3>
<p>The general form for Wright's Fst equation is:&nbsp;</p>
<p><img class="equation_image" title="\frac{H_T-H_S}{H_T}" src="https://clemson.instructure.com/equation_images/%255Cfrac%257BH_T-H_S%257D%257BH_T%257D" alt="LaTeX: \frac{H_T-H_S}{H_T}" data-equation-content="\frac{H_T-H_S}{H_T}" data-mathml="&lt;math xmlns=&quot;http://www.w3.org/1998/Math/MathML&quot;&gt;
  &lt;mfrac&gt;
    &lt;mrow&gt;
      &lt;msub&gt;
        &lt;mi&gt;H&lt;/mi&gt;
        &lt;mi&gt;T&lt;/mi&gt;
      &lt;/msub&gt;
      &lt;mo&gt;&amp;#x2212;&lt;!-- &minus; --&gt;&lt;/mo&gt;
      &lt;msub&gt;
        &lt;mi&gt;H&lt;/mi&gt;
        &lt;mi&gt;S&lt;/mi&gt;
      &lt;/msub&gt;
    &lt;/mrow&gt;
    &lt;msub&gt;
      &lt;mi&gt;H&lt;/mi&gt;
      &lt;mi&gt;T&lt;/mi&gt;
    &lt;/msub&gt;
  &lt;/mfrac&gt;
&lt;/math&gt;" />&nbsp;</p>
<p><img class="equation_image" title="H_T=H_{exp_T}" src="https://clemson.instructure.com/equation_images/H_T%253DH_%257Bexp_T%257D" alt="LaTeX: H_T=H_{exp_T}" data-equation-content="H_T=H_{exp_T}" data-mathml="&lt;math xmlns=&quot;http://www.w3.org/1998/Math/MathML&quot;&gt;
  &lt;msub&gt;
    &lt;mi&gt;H&lt;/mi&gt;
    &lt;mi&gt;T&lt;/mi&gt;
  &lt;/msub&gt;
  &lt;mo&gt;=&lt;/mo&gt;
  &lt;msub&gt;
    &lt;mi&gt;H&lt;/mi&gt;
    &lt;mrow class=&quot;MJX-TeXAtom-ORD&quot;&gt;
      &lt;mi&gt;e&lt;/mi&gt;
      &lt;mi&gt;x&lt;/mi&gt;
      &lt;msub&gt;
        &lt;mi&gt;p&lt;/mi&gt;
        &lt;mi&gt;T&lt;/mi&gt;
      &lt;/msub&gt;
    &lt;/mrow&gt;
  &lt;/msub&gt;
&lt;/math&gt;" />&nbsp;, i.e.&nbsp;the expected heterozygosity for the total population (so we already have this value calculated).</p>
<p>H<sub>s</sub> is a weighted average of the expected heterozygosities for all sub-populations. It is calculated like this:</p>
<p><img class="equation_image" title="H_S=\sum^k_{i=1}\frac{n_i}{n}H_{exp_i}" src="https://clemson.instructure.com/equation_images/H_S%253D%255Csum%255Ek_%257Bi%253D1%257D%255Cfrac%257Bn_i%257D%257Bn%257DH_%257Bexp_i%257D" alt="LaTeX: H_S=\sum^k_{i=1}\frac{n_i}{n}H_{exp_i}" data-equation-content="H_S=\sum^k_{i=1}\frac{n_i}{n}H_{exp_i}" data-mathml="&lt;math xmlns=&quot;http://www.w3.org/1998/Math/MathML&quot;&gt;
  &lt;msub&gt;
    &lt;mi&gt;H&lt;/mi&gt;
    &lt;mi&gt;S&lt;/mi&gt;
  &lt;/msub&gt;
  &lt;mo&gt;=&lt;/mo&gt;
  &lt;munderover&gt;
    &lt;mo&gt;&amp;#x2211;&lt;!-- &sum; --&gt;&lt;/mo&gt;
    &lt;mrow class=&quot;MJX-TeXAtom-ORD&quot;&gt;
      &lt;mi&gt;i&lt;/mi&gt;
      &lt;mo&gt;=&lt;/mo&gt;
      &lt;mn&gt;1&lt;/mn&gt;
    &lt;/mrow&gt;
    &lt;mi&gt;k&lt;/mi&gt;
  &lt;/munderover&gt;
  &lt;mfrac&gt;
    &lt;msub&gt;
      &lt;mi&gt;n&lt;/mi&gt;
      &lt;mi&gt;i&lt;/mi&gt;
    &lt;/msub&gt;
    &lt;mi&gt;n&lt;/mi&gt;
  &lt;/mfrac&gt;
  &lt;msub&gt;
    &lt;mi&gt;H&lt;/mi&gt;
    &lt;mrow class=&quot;MJX-TeXAtom-ORD&quot;&gt;
      &lt;mi&gt;e&lt;/mi&gt;
      &lt;mi&gt;x&lt;/mi&gt;
      &lt;msub&gt;
        &lt;mi&gt;p&lt;/mi&gt;
        &lt;mi&gt;i&lt;/mi&gt;
      &lt;/msub&gt;
    &lt;/mrow&gt;
  &lt;/msub&gt;
&lt;/math&gt;" />, where&nbsp;<em>k =&nbsp;</em><em>the num. of sub-pops,&nbsp;</em><em>n</em><em> = the total population size,&nbsp;</em><em>n<em><sub>i</sub></em><em> = the</em> size of the i<sup>th</sup> sub-population, and H<sub>exp</sub><sub>i</sub> = the expected heterozygosity of the i<sup>th</sup> sub-population.</em></p>
<p>For our data set today, where we have 2 sub-populations, this evaluates to:</p>
<p><img class="equation_image" title="H_S=\left[\frac{n_1}{n}\cdot H_{exp_1}\right]+\left[\frac{n_2}{n}\cdot H_{exp_2}\right]" src="https://clemson.instructure.com/equation_images/H_S%253D%255Cleft%255B%255Cfrac%257Bn_1%257D%257Bn%257D%255Ccdot%2520H_%257Bexp_1%257D%255Cright%255D%2B%255Cleft%255B%255Cfrac%257Bn_2%257D%257Bn%257D%255Ccdot%2520H_%257Bexp_2%257D%255Cright%255D" alt="LaTeX: H_S=\left[\frac{n_1}{n}\cdot H_{exp_1}\right]+\left[\frac{n_2}{n}\cdot H_{exp_2}\right]" data-equation-content="H_S=\left[\frac{n_1}{n}\cdot H_{exp_1}\right]+\left[\frac{n_2}{n}\cdot H_{exp_2}\right]" data-mathml="&lt;math xmlns=&quot;http://www.w3.org/1998/Math/MathML&quot;&gt;
  &lt;msub&gt;
    &lt;mi&gt;H&lt;/mi&gt;
    &lt;mi&gt;S&lt;/mi&gt;
  &lt;/msub&gt;
  &lt;mo&gt;=&lt;/mo&gt;
  &lt;mrow&gt;
    &lt;mo&gt;[&lt;/mo&gt;
    &lt;mrow&gt;
      &lt;mfrac&gt;
        &lt;msub&gt;
          &lt;mi&gt;n&lt;/mi&gt;
          &lt;mn&gt;1&lt;/mn&gt;
        &lt;/msub&gt;
        &lt;mi&gt;n&lt;/mi&gt;
      &lt;/mfrac&gt;
      &lt;mo&gt;&amp;#x22C5;&lt;!-- &sdot; --&gt;&lt;/mo&gt;
      &lt;msub&gt;
        &lt;mi&gt;H&lt;/mi&gt;
        &lt;mrow class=&quot;MJX-TeXAtom-ORD&quot;&gt;
          &lt;mi&gt;e&lt;/mi&gt;
          &lt;mi&gt;x&lt;/mi&gt;
          &lt;msub&gt;
            &lt;mi&gt;p&lt;/mi&gt;
            &lt;mn&gt;1&lt;/mn&gt;
          &lt;/msub&gt;
        &lt;/mrow&gt;
      &lt;/msub&gt;
    &lt;/mrow&gt;
    &lt;mo&gt;]&lt;/mo&gt;
  &lt;/mrow&gt;
  &lt;mo&gt;+&lt;/mo&gt;
  &lt;mrow&gt;
    &lt;mo&gt;[&lt;/mo&gt;
    &lt;mrow&gt;
      &lt;mfrac&gt;
        &lt;msub&gt;
          &lt;mi&gt;n&lt;/mi&gt;
          &lt;mn&gt;2&lt;/mn&gt;
        &lt;/msub&gt;
        &lt;mi&gt;n&lt;/mi&gt;
      &lt;/mfrac&gt;
      &lt;mo&gt;&amp;#x22C5;&lt;!-- &sdot; --&gt;&lt;/mo&gt;
      &lt;msub&gt;
        &lt;mi&gt;H&lt;/mi&gt;
        &lt;mrow class=&quot;MJX-TeXAtom-ORD&quot;&gt;
          &lt;mi&gt;e&lt;/mi&gt;
          &lt;mi&gt;x&lt;/mi&gt;
          &lt;msub&gt;
            &lt;mi&gt;p&lt;/mi&gt;
            &lt;mn&gt;2&lt;/mn&gt;
          &lt;/msub&gt;
        &lt;/mrow&gt;
      &lt;/msub&gt;
    &lt;/mrow&gt;
    &lt;mo&gt;]&lt;/mo&gt;
  &lt;/mrow&gt;
&lt;/math&gt;" /></p>
<p>Once you have calculated Hs, it should be very straightforward to calculate Fst.</p>
<h3 id="3plottheresults">3. Plot the Results</h3>
<h5 id="distribution">First, let's look at the frequency distribution of all of the Fst values, using R's built-in histogram function:</h5>
<pre><code>hist(my.results$Fst, xlim=c(0,1), breaks=50,xlab=expression(F[ST]), main="Histogram of FST", col="lightblue")
</code></pre>
<p>The arguments that I'm using in this function are:</p>
<ol>
<li>The first argument is always the data that you want to plot, in this case <code>my.results$Fst</code>.</li>
<li><strong>xlim</strong> sets the x-axis range to plot; if we didn't specify this, R would just pick the range automatically based on the data.</li>
<li><strong>breaks</strong> tells R what size bins to use for the histogram; again, if we didn't specify this parameter, R would pick it for us automatically.</li>
<li><strong>xlab</strong> Creates a label for the x-axis. I'm using the additional function <code>expression()</code> here to generate a label with a sub-script</li>
<li><strong>main</strong> creates the title for the plot</li>
<li><strong>col</strong> sets the plot color. Here is a <a href="http://www.stat.columbia.edu/~tzheng/files/Rcolor.pdf"><img src="http://www.stat.columbia.edu/~tzheng/files/Rcolor.pdf" alt="handy pdf" /></a> showing all of the built-in color options in R. You can get even more with extra packages.</li>
</ol>
<h5 id="genomeplot">Next, to look at how the Fst values are distributed spatial along the genome, we can simply make a scatterplot where Fst is a function of position.</h5>
<pre><code>plot(my.data$POS, my.results$Fst, xlab="Chromosomal Position (bp)", ylab=expression(F[ST]), ylim=c(0,1), pch=20, main="Spatial Distribution of Fst", col="hotpink3")
</code></pre>
<p>Here, the plot arguments are:</p>
<ol>
<li>The first value should always be the <em>x</em> values; in this case the positions</li>
<li>The second argument should always be the <em>y</em> values, which here is Fst.</li>
<li><strong>xlab</strong> does the same thing here as it did in <code>hist()</code> and <strong>ylab</strong> makes a label for the y-axis.</li>
<li><strong>ylim</strong> sets the limits for the y-axis.</li>
<li><strong>pch</strong> specifies which plotting symbol to use. I specify #20 for closed circles.</li>
<li><strong>main</strong> again gives the plot a title, and <strong>col</strong> again specifies the color.</li>
</ol>
<p>What does looking at the histogram tell you about Fst? What does looking at the genomewide plot tell you?</p>
<h3 id="extrapcatolookatpopulationstructure">EXTRA: PCA to look at Population Structure</h3>
<p>If we hadn't known before starting our analysis that we had 2 sub-groups in the data, then we might have wanted to start with a more general analysis to detect population structure.</p>
<p>Principal Components Analysis (PCA) is one method of doing this. You can find a very basic overview of PCA <a href="https://georgemdallas.wordpress.com/2013/10/30/principal-component-analysis-4-dummies-eigenvectors-eigenvalues-and-dimension-reduction/"><img src="https://georgemdallas.wordpress.com/2013/10/30/principal-component-analysis-4-dummies-eigenvectors-eigenvalues-and-dimension-reduction/" alt="here" /></a> if you're not familiar with it.</p>
<h5 id="partaconvertgenotypestoamatrixofnumericvalues"><strong>Part A: Convert Genotypes to a Matrix of Numeric Values</strong></h5>
<p>R can only perform PCA on numerical values, so we need to convert our genotypes to numbers that R can handle.</p>
<p>First, set up an empty <strong>matrix</strong> with the same number of rows as <code>my.data</code> and a column for every sample:</p>
<pre><code>numericGen = matrix(NA, nrow=nrow(my.data), ncol=(ncol(my.data) - 9))
colnames(numericGen) = colnames(my.data)[10:ncol(my.data)]
</code></pre>
<p>Now, you need to write a <code>for()</code> loop that goes through &nbsp;<code>my.data</code> and replaces the genotype strings with numbers, then saves the new numeric strings in the matrix.</p>
<p>This is a lot like Part 3 of the <a
	href="/gen8900/exercises/vcf.html">VCF Exercises</a> where we converted the VCF genotypes to the diploid nucleotides. The only difference is now we convert them to the following numbers:</p>
<p>"0/0" = 1</p>
<p>"0/1" = 2</p>
<p>"1/1" = 3</p>
<p>"./." = 0</p>
<p><strong>Note</strong>: Normally, you would <em>not</em> want to replace missing data with 0. For the purposes of this exercise, we are doing it as a crude but quick way to avoid the problems R has with missing data values in PCA. In a real analysis, you would want to use some sort of imputation method specific to genetic data to deal with these values.</p>
<h5 id="partbrunpcaandplottheresults"><strong>Part B: Run PCA and plot the results</strong></h5>
<p>For PCA in R, we need to transpose our matrix so that the samples become the rows, and the SNPs become the columns. This is easily done with the <code>t()</code> function:</p>
<pre><code>numGen.trans = t(numericGen)
</code></pre>
<p>Now, run PCA with the <code>prcomp</code> function:</p>
<pre><code>pca.results = prcomp(numGen.trans, scale=FALSE)
</code></pre>
<p>There a few things we can look at in <code>pca.results</code>. First, if you type <code>summary(pca.results</code> you will get a table of information about the principal components, including how much variance is explained by each.</p>
<p>If you type <code>plot(pca.results)</code> you will see the variance explained by each PC.</p>
<p>Finally, the plot that we are most interested in is a plot of the first two PCs, which will show us if our data is clustered at all.</p>
<p>To get our plot points to be colored by species, first run this:</p>
<pre><code>plot.col = gsub("^FU_.*", "darkturquoise", gsub("^MA_.*", "darkviolet", rownames(numGen.trans)))
</code></pre>
<p>Now, create the plot:</p>
<pre><code>plot(pca.results$x, col=plot.col, pch=20, main="PCA of Finch Data")
legend(18, -1, c("Fuliginosa", "Magnirostris"), col=c("darkturquoise", "darkviolet"), pch=c(20,20), cex=0.75)
</code></pre>
<p>What can you tell about population structure from the plot? How
	does the information in the 3 plots we generated today
differ?</p>



<a href="/gen8900/r-code/4_FST_solutions.R" download="4_FST_solutions.R">R code solution for Fst Exercises</a>

	  </div>
	</div>
</div>
</section>

<!-- Footer -->
    <footer class="py-5 bg-dark">
      <div class="container">
	<a class="navbar-brand" href="#top">Back to Top</a>
        <p class="m-0 text-center text-white">Copyright &copy; LCooper CompGen-8900  2017</p>
      </div>
      <!-- /.container -->
    </footer>

    <!-- Bootstrap core JavaScript -->
    <script src="vendor/jquery/jquery.min.js"></script>
    <script src="vendor/popper/popper.min.js"></script>
      <script src="vendor/bootstrap/js/bootstrap.min.js"></script>
      
</body>
</html>