<!DOCTYPE html>
<html lang="en">

  <head>

    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
    <meta name="description" content="">
    <meta name="author" content="">

    <title>Syllabus</title>

    <!-- Bootstrap core CSS -->
    <link href="../vendor/bootstrap/css/bootstrap.min.css" rel="stylesheet">

    <!-- Custom styles for this template -->
    <link href="../css/1-col-portfolio.css" rel="stylesheet">

  </head>

  <body>
    <!-- Navigation -->
    <nav class="navbar navbar-expand-lg navbar-dark bg-dark fixed-top">
      <div class="container">
        <a class="navbar-brand" href="#">GEN-8900 Course Syllabus</a>
        <button class="navbar-toggler" type="button" data-toggle="collapse" data-target="#navbarResponsive" aria-controls="navbarResponsive" aria-expanded="false" aria-label="Toggle navigation">
          <span class="navbar-toggler-icon"></span>
        </button>
        <div class="collapse navbar-collapse" id="navbarResponsive">
          <ul class="navbar-nav ml-auto">
            <li class="nav-item active">
              <a class="nav-link" href="../compGen8900.html">Course Home Page
                <span class="sr-only">(current)</span>
              </a>
            </li>
	   </ul>
	 </div>
	</div>
	</nav>

    <!-- Content -->
    <div class="container">
      <div class="row">
	<div class="col-md-12">
	  <h2>Computational Genomics Workshop:</h2>
	  <h3>R coding for the analysis of SNP data and applied population genetics</h3>
	  <p><strong>Description:</strong></p>
	  <p>High-throughput sequencing technology has made it possible to obtain large scale genetic data sets for almost any organism, creating a need for computational tools and skill sets to process these data.&nbsp; While the bioinformatics workflows for processing raw data into SNPs are typically well delineated, the path for analyzing and interpreting the resulting SNP data set can be less clear.&nbsp; Classical population genetics theory is a rich field with many existing statistics and mathematical models, all created to draw biological inferences from nothing more than the patterns of observed gene and allele frequencies.&nbsp; Since these frequencies can be calculated from any SNP data set, this makes them highly applicable to current research.&nbsp; In this workshop, students will learn about statistics that test the neutral theory, and then apply the test to a sample SNP data set.&nbsp; Throughout this process, students will learn to write their own R code for each analysis, and emphasis will be placed on algorithm design with scalability for large data sets in mind.&nbsp;</p>
	  <p>&nbsp;</p>
	  <p><strong>Goals:</strong></p>
	  <ol>
	    <li>Achieve a thorough understanding of each statistic or model, especially in terms of neutral theory expectations, and interpretations of deviations from the neutral expectation. Emphasis will also be placed on knowing when a certain model may or may not be appropriate for a particular data set or scenario.</li>
	    <li>Make the connection between the table of genotypes obtained in a VCF file (standard next-gen pipeline output) and a theoretical equation that comes from a textbook or paper. By writing their own code each week, students will get a feel for the best ways to set up algorithms by dividing tasks into re-usable functions, parsing difficult file structures, and creating useful output.&nbsp;</li>
	    <li>Learn (or improve) R programming skills with hands-on practice. For each exercise, we will focus on writing code from scratch (rather than using pre-existing packages), so that students will become familiar with common coding structures such as functions and loops.&nbsp; These skills are applicable to all coding languages, rather than being R specific.</li>
	  </ol>
	  <p>&nbsp;</p>
	  <p><strong>Prerequisites:</strong></p>
	  <ul>
	    <li>GEN 3000 or an equivalent undergrad genetics course from another institution. Students must have a working knowledge of the basic concepts of genetics, including Mendelian inheritance and the molecular basis of heredity.&nbsp; Prior knowledge of population genetics is NOT required.</li>
	    <li>ALL LEVELS OF PROGRAMMING EXPERIENCE ARE WELCOME! This course is designed for both biologists who want to learn bioinformatics, and for people with programming experience who want to learn more about biology or population genomics.</li>
	    <li>In the previous semester, some students who were completely new to R found it helpful to also go through an online tutorial of R to get more comfortable with it in the beginning. They recommended this one: http://nathanieldphillips.com/thepiratesguidetor/&nbsp;</li>
	  </ul>
	  <p>&nbsp;</p>
	  <p><strong>General Structure:</strong></p>
	  <p>2.5 credits; Grades assigned based on participation (70%) and final project (30%)</p>
	  <p>Meet 1 time per week, for 2.5 hours<sup>*</sup></p>
	  <p><sup>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; </sup>
	    <strong>First 30-45 minutes</strong>: Review results or questions from the previous week.&nbsp; Then, discuss one key test statistic/concept; go over in detail the meaning, derivation, assumptions, and expectations of the central model or statistic for the week.</p>
	  <p><strong>1.5 hours</strong>: Briefly discuss algorithm for implementing the model in a VCF file, then everyone will work on their code while interacting with the instructor and the other students for assistance and feedback<sup>*</sup>.</p>	  <p><em><sup>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;*</sup>
	    </em><em>The exact amount of time devoted to lecture vs. in-class programming will change from week to week, depending on what the topic is.&nbsp; Additionally, some weeks the in-class exercise may not take the full amount of time, while other exercises might span into the next week of class.</em></p>
	  <p><strong>&nbsp;</strong></p>
	  <p><strong>Meeting Time:</strong></p>
	  <p>TUESDAYS 2:00PM &ndash; 4:30PM</p>
	  <p><strong>&nbsp;</strong></p>
	  <p><strong>Week 1: Introduction to Course and R basics</strong></p>
	  <ul>
	    <li>Review the syllabus and course structure.</li>
	    <li>Brief introductions.</li>
	    <li>R-tutorial: work through creating and manipulating variables, common data structures, performing built-in functions, and writing loops.</li>
	    <li>Getting help: finding and reading R documentation.</li>
	    <li>Exercises: Using R to perform some simple tasks, including creating a table and writing a small loop to perform some calculation on each row.</li>
	  </ul>
	  <p><strong>Week 2: The Variant Call Format (VCF)</strong></p>
	  <ul>
	    <li>Discuss the major concepts involved in next-gen sequencing, and review the standard raw data pipelines that are used to call SNPs and produce VCF files.</li>
	    <li>Look in detail at the VCF format, and learn how to read this format into R and manipulate the resulting table.</li>
	    <li>Exercise: Read a sample VCF file into R, and perform some simple functions to extract, summarize, and re-code the SNP data.</li>
	  </ul>
	  <p><strong><em>&nbsp;</em></strong></p>
	  <p><strong><em>Single-Site Tests</em></strong></p>
	  <p><strong>Week 3: Hardy-Weinberg Equilibrium</strong></p>
	  <ul>
	    <li>Review principle of HWE; its usefulness as a diagnostic test in next-gen data; calculating observed and expected frequencies; thinking about when you do and don&rsquo;t expect deviations.</li>
	    <li>Use Fisher&rsquo;s Exact test to find statistically significant deviation.</li>
	    <li>Exercise: Write a script to calculate observed and expected genotype frequencies at each site in a sample VCF file, then extract sites that show too much deviation</li>
	  </ul>
	  <p><strong>Week 4: F<sub>ST</sub> </strong></p>
	  <ul>
	    <li>Discussion of population structure, what causes it, and how F<sub>ST</sub> is used to measure it.</li>
	    <li>Equations and assumptions for F<sub>ST</sub>; effects of population size on genetic drift; relationship between F<sub>ST</sub> and migration; review of alternative F<sub>ST</sub> estimators (and when to use them).</li>
	    <li>Exercise: write code to calculate single site F<sub>ST</sub> (using Wright&rsquo;s original equation based on single site frequencies) between 2 populations in a sample VCF file; plot the distribution of F<sub>ST</sub>.</li>
</ul>
	  <p><strong><em>Pairwise Tests</em></strong></p>
	  <p><strong>Week 5: Linkage Disequilibrium pt. 1 </strong></p>
	  <ul>
	    <li>Definition of LD and underlying causes. Estimators of LD, and using haplotypes versus inferred haplotypes.</li>
	    <li>Exercise: write a script to find 2-locus haplotype frequencies (in addition to single site allele frequency), and calculate the LD estimators D and <em>r<sup>2</sup> </em>between every PAIR of SNPs in a very small (test case) VCF file. Use previous functions for genotype and allele frequencies, but create a new nested loop structure that will get site pairs.&nbsp; Save functions and code to run on a larger file next week.</li>
	  </ul>
	  <p><strong>Week 6: Linkage Disequilibrium pt. 2</strong></p>
	  <ul>
	    <li>The interpretation and meaning of <em>LD decay</em>. How to estimate the recombination rate (rho) based on the mathematical relationship between LD and recombination.</li>
	    <li>Exercise: Using the code from last week, calculate r<sup>2</sup> in a full 1000 row VCF file and plot the decay of LD with distance. Fit a non-linear model to estimate recombination from the rate of decay.</li>
	  </ul>
	  <p><strong>Fall Break</strong></p>
	  <p><strong>Review and Catch-Up</strong></p>
	  <ul>
	    <li>Review the algorithms for single site tests and for calculating r<sup>2</sup> and plotting the decay of LD with distance.</li>
	    <li>Review of loop structures and behaviors.</li>
	    <li>R exercise: finish up solutions from previous weeks before fall break; create a new type of plot (the LD heatmap).</li>
	  </ul>
	  <p><strong><em>Multi-Locus Tests (Combined sites and sliding windows)</em></strong></p>
	  <p><span><strong>Week 9: Allele Frequency Spectrum and Mutation Rates</strong></span></p>
	  <ul>
	    <li>A brief introduction to coalescent theory.</li>
	    <li>Neutral (coalescent) theory expectations of allele frequency distributions; selective and demographic forces causing deviations from neutral; interpretation of single gene vs. genome-wide spectra.</li>
	    <li>Define Theta and its estimators; discuss Watterson&rsquo;s theta vs. Pi; how to calculate each; the different expectations of each, and how this relates back to the Allele Frequency Spectrum.</li>
	    <li>Exercise: For a VCF file with SNPs from a single locus; plot the observed allele frequency spectrum using R's histogram function, and use a given function to simulate the curve for the neutral expectation.&nbsp; Then calculate both estimators of theta, and&nbsp;see if they differ in the way that you would expect based on your frequency spectrun plots.</li>
	    <li>Both estimators will require some new functions for the calculations; the Pi estimator in particular will require a new strategy of comparing columns in a VCF rather than rows; Watterson&rsquo;s theta should be pretty straightforward.</li>
	  </ul>
	  <p><span style="background-color: #ffff00;"><strong>Deadline to Select Topic for Final Project!&nbsp;</strong></span></p>
	  <p><strong>Week 10: Tajima's D</strong></p>
	  <ul>
	    <li>Review the different types of selection, and their predicted effects on the allele frequency spectrum.</li>
	    <li>Testing for selection using the difference in estimates of theta; how to calculate D; different types of selection predicted by negative vs. positive D. Discuss extensions of D (e.g. Fay and Wu&rsquo;s H).</li>
	    <li>Exercise: Calculate and compare values of Tajima's D for several genome regions provided in a VCF file.&nbsp; Determine which gene/region might be under selection.</li>
	  </ul>
	  <p><strong>Week 11: Sliding Window Tests with Missing Data</strong></p>
	  <ul>
	    <li>Why and how do you get gaps or missing data in next-gen sequencing? What are the implications of missing data in terms of 1) What tests you should perform on your data, and 2) what programs you should run to get the most information possible?</li>
	    <li>Review statistics that are often used in sliding window tests (Pi, Dxy, multi-site Fst, Tajima's D, etc.)</li>
	    <li>Learn about statistics related to Tajima's D: Fu and Li's F and D, Fay and Wu's H</li>
	    <li>Exercise: perform a sliding window test to plot compare&nbsp;a pair of statistics on a sample VCF file.</li>
	  </ul>
	  <p>&nbsp;<strong>Week 12: McDonald-Kreitman Test</strong></p>
	  <ul>
	    <li>Review the theory underlying the MK test to find sites under positive selection.</li>
	    <li>>Discussion of gene annotation: what this means, and what kind of programs give you this information.</li>
	    <li>Exercise: using an already annotated sample VCF file, perform the MK test to find sites under selection.</li>
	  </ul>

	  <p><strong><em>Scaling Up</em></strong></p>
	  <p><span style="color: #333333;"><strong>Week 13: Running R on the cluster </strong></span></p>
	  <ul>
	    <li>Brief overview of the linux command line environment (navigating directories, creating, deleting, splitting, concatenating, and moving files).</li>
	    <li>Cluster basics: logging in, transferring files, using interactive nodes and submitting job scripts.</li>
	    <li>Running R from the command line; writing output for data and figures to files; making code more flexible with command line arguments.</li>
	    <li>In-class time to work on projects and get feedback as needed.</li>
	  </ul>
	  <p><strong>Last Day of Class: Project Presentations</strong></p>
	  <ul>
	    <li>10 minute presentations by each person on the results of their project.</li>
	  </ul>
	</div>
      </div>
    </div>
  </body>
</html>
